### ISSUE TITLE
Observed nil flow from source to dereference point: result 0 of `Do()` accessed field `Body`

### ISSUE THREAD

**Issue by skhaz** [uber-go/nilaway#126]
I am trying the new Uber's tool, NilAway.

In this code

```
resp, err := client.Do(req)
if err != nil {
	return Runtime{}, fmt.Errorf("http get error: %w", err)
}
defer resp.Body.Close()

body, err := io.ReadAll(resp.Body)
if err != nil {
	return Runtime{}, fmt.Errorf("read all error: %w", err)
}
```

I get the error

```
Error: /home/runner/work/***/***/main.go:119:26: error: Potential nil panic detected. Observed nil flow from source to dereference point: 
Error: 	-> http/client.go:644:12: unassigned variable `resp` returned from `do()` in position 0
Error: 	-> http/client.go:590:9: result 0 of `do()` returned from `Do()` in position 0
Error: 	-> ***/main.go:119:26: result 0 of `Do()` accessed field `Body`
```

If I understand correctly, the tool wants me to deal with this potential nil. This way.

```
if err != nil || resp == nil {
```

But that's somewhat redundant, since if the error is nil, the response will be too.

What do you think?

**Comment by yuxincs** [comment]
> But that's somewhat redundant, since if the error is nil, the response will be too.

Yes you're right. In general NilAway handles such cases, where it would try to reason that if the error-returning functions are _only_ returning `<NONNIL, NIL>` and `<NIL, NONNIL>`, then these mutually exclusivity is considered and won't trigger an error if you are checking `if err != nil {...}`.

Here the case is more interesting. Correct me if anyone is more familiar with the code in `http/client.go`:

(1) Local vars `reqs []*Request` and `resp *Response` are uninitialized here: 

https://github.com/golang/go/blob/37971b31dfde1b9c40f6d42484e528c17450a9ae/src/net/http/client.go#L607-L617

(2) we then enter the for loop below, with empty `reqs` and `nil` `resp`

https://github.com/golang/go/blob/37971b31dfde1b9c40f6d42484e528c17450a9ae/src/net/http/client.go#L635-L645

Now that the only reason it doesn't result in a panic is because anything under `if len(reqs) > 0 {` is dead code and would never be executed.

Am I missing something? (in the meantime we could probably enhance NilAway to understand length checks better)

**Comment by hamdouni** [comment]
there is also an implicit contract that if err != nil then resp can not be nil (exactly, that resp.Body is not nil)

**Comment by karelbilek** [comment]
Is there a way to model the standard library/tell nilaway about the contract explicitly?

**Comment by karelbilek** [comment]
I don't want to be the one doing +1 but this bug still being present stops me from using nilaway :( 

can I do some kind of "external" annotation to standard library?
